### ISSUE TITLE
(?{...}) pushes onto wrong variable?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5361]
Migrated from [rt.perl.org#9010](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9010) (status was 'resolved')

Searchable as RT9010$

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



I don't know what the subject of this message should be\.  The report
concerns this program&#8203;:

&nbsp;    for \(1\.\.2\) \{
&nbsp;      my @&#8203;s;
&nbsp;      /\(?\{push @&#8203;s\, "x"\}\)/;
&nbsp;      print "$\_&#8203;: @&#8203;s\\n";
&nbsp;    \}

I believe the output should be

&nbsp;        1&#8203;: x
&nbsp;        2&#8203;: x

However\, it is

&nbsp;        1&#8203;: x
&nbsp;        2&#8203;: 

The problem persists in bleadperl\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by root at Sat Dec 29 11:54:59 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.2-2, archname=i586-linux
    uname='linux plover.com 2.4.2-2 #1 sun apr 8 19:37:14 edt 2001 i586 unknown '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-81)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i586-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/mjd
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11R6:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/teTeX/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Apr 24\, mjd@&#8203;plover\.com said&#8203;:

>I don't know what the subject of this message should be\.  The report
>concerns this program&#8203;:
>
>    for \(1\.\.2\) \{
>      my @&#8203;s;
>      /\(?\{push @&#8203;s\, "x"\}\)/;
>      print "$\_&#8203;: @&#8203;s\\n";
>    \}
>
>I believe the output should be
>
>        1&#8203;: x
>        2&#8203;: x
>
>However\, it is
>
>        1&#8203;: x
>        2&#8203;: 
>
>The problem persists in bleadperl\.

The problem is that the regex is only compiled once\, and so the code\-eval
assertion \(?\{ \}\) ends up referring to the first my\(@&#8203;s\) only\.  If you can
force the regex to recompile\, the problem disappears\.

This can be done \(at least\) two ways&#8203;:

&nbsp;  for \(1\, 2\) \{
&nbsp;    my @&#8203;s;
&nbsp;    eval '/\(?\{ push @&#8203;s\, "x" \}\)/';
&nbsp;    print "\<@&#8203;s>\\n";
&nbsp;  \}

or

&nbsp;  use re 'eval';
&nbsp;  for \(1\, 2\) \{
&nbsp;    my @&#8203;s;
&nbsp;    /\(?\{ push @&#8203;s\, "x" \}\)|$\_/;
&nbsp;    print "\<@&#8203;s>\\n";
&nbsp;  \}

The problem is that since the regex only gets compiled once\, so too the
code blocks are only compiled once\.  Perhaps we need a flag \(or the
default\) to be that code blocks are recompiled?

Here is proof that what I'm saying is true&#8203;:

&nbsp;  for \(1\, 2\) \{
&nbsp;    my @&#8203;s;
&nbsp;    /\(?\{ push @&#8203;s\, "x" \}\)/;
&nbsp;    push @&#8203;t\, \\@&#8203;s;
&nbsp;  \}
&nbsp;  print "\[@&#8203;$\_\]\\n" for @&#8203;t;  \# \[x x\] \[\]

I actually ran into this problem once before\.

\-\- 
Jeff "japhy" Pinyan      japhy@&#8203;pobox\.com      http&#8203;://www\.pobox\.com/~japhy/
RPI Acacia brother \#734   http&#8203;://www\.perlmonks\.org/   http&#8203;://www\.cpan\.org/
\*\* Look for "Regular Expressions in Perl" published by Manning\, in 2002 \*\*
\<stu> what does y/// stand for?  \<tenderpuss> why\, yansliterate of course\.
\[  I'm looking for programming work\.  If you like my work\, let me know\.  \]

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.04\.24 23&#8203;:07 Jeff 'japhy' Pinyan wrote&#8203;:
> 
> The problem is that the regex is only compiled once\, and so the code\-eval
> assertion \(?\{ \}\) ends up referring to the first my\(@&#8203;s\) only\.  If you can
> force the regex to recompile\, the problem disappears\.

In other words\, this should at least trigger a warning à la
"Variable "%s" will not stay shared"\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In article \<Pine\.GSO\.4\.21\.0204241658580\.18750\-100000@&#8203;crusoe\.crusoe\.net>\,
&nbsp;    Jeff 'japhy' Pinyan \<jeffp@&#8203;crusoe\.net> writes&#8203;:
> On Apr 24\, mjd@&#8203;plover\.com said&#8203;:
> 
>>I don't know what the subject of this message should be\.  The report
>>concerns this program&#8203;:
>>
>>    for \(1\.\.2\) \{
>>      my @&#8203;s;
>>      /\(?\{push @&#8203;s\, "x"\}\)/;
>>      print "$\_&#8203;: @&#8203;s\\n";
>>    \}
>>
> 
> The problem is that the regex is only compiled once\, and so the code\-eval
> assertion \(?\{ \}\) ends up referring to the first my\(@&#8203;s\) only\.  If you can
> force the regex to recompile\, the problem disappears\.
> 

I don't really understand this explanation\.
If i have code in a normal for loop\, it's also compiled only once and
still can use a lexical variable and access a different value each loop\.

So i assume you really mean "regex compile"\, not "code compile"\. But why
would the regex compile somehow influence how code fragments work ? I'd
still expect @&#8203;s to be accessed through the pad and so influence the
"right" one\.

**Comment by p5pRT** [comment]
### From @tamias

On Wed\, Apr 24\, 2002 at 10&#8203;:20&#8203;:21PM \+0000\, Ton Hospel wrote&#8203;:
> In article \<Pine\.GSO\.4\.21\.0204241658580\.18750\-100000@&#8203;crusoe\.crusoe\.net>\,
>     Jeff 'japhy' Pinyan \<jeffp@&#8203;crusoe\.net> writes&#8203;:
> > On Apr 24\, mjd@&#8203;plover\.com said&#8203;:
> > 
> >>I don't know what the subject of this message should be\.  The report
> >>concerns this program&#8203;:
> >>
> >>    for \(1\.\.2\) \{
> >>      my @&#8203;s;
> >>      /\(?\{push @&#8203;s\, "x"\}\)/;
> >>      print "$\_&#8203;: @&#8203;s\\n";
> >>    \}
> >>
> > 
> > The problem is that the regex is only compiled once\, and so the code\-eval
> > assertion \(?\{ \}\) ends up referring to the first my\(@&#8203;s\) only\.  If you can
> > force the regex to recompile\, the problem disappears\.
> > 
> 
> I don't really understand this explanation\.
> If i have code in a normal for loop\, it's also compiled only once and
> still can use a lexical variable and access a different value each loop\.
> 
> So i assume you really mean "regex compile"\, not "code compile"\. But why
> would the regex compile somehow influence how code fragments work ? I'd
> still expect @&#8203;s to be accessed through the pad and so influence the
> "right" one\.


It's the same behavior as nested named subs&#8203;:

\#\!perl \-w

sub foo \{
&nbsp;  my @&#8203;s;

&nbsp;  sub bar \{
&nbsp;    push @&#8203;s\, 'x';
&nbsp;  \}

&nbsp;  bar\(\);
&nbsp;  print "$\_\[0\]&#8203;: @&#8203;s\\n";
\}

foo\($\_\) for 1\, 2;
\_\_END\_\_

**Comment by p5pRT** [comment]
### From @mjdominus


> The problem is that the regex is only compiled once\, and so the code\-eval
> assertion \(?\{ \}\) ends up referring to the first my\(@&#8203;s\) only\.  

I don't understand what 'compiled only once' has to do with it\.  In

&nbsp;        for \(1\.\.10\) \{
&nbsp;          my $a;
&nbsp;          /\(?\{ $a\+=1 \}\)/;
&nbsp;          print $a;
&nbsp;        \}

The "print $a" line is compiled only once\, and nevertheless manages to
get the right lexical variable every time through\.  The compiled code
in the regex should refer to $a in exactly the same way that the
"print $a" does\.

It seems to me that this is a bug\, plain and simple\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In article \<20020424182628\.A194890@&#8203;linguist\.thayer\.dartmouth\.edu>\,
&nbsp;    Ronald J Kimball \<rjk@&#8203;linguist\.Thayer\.dartmouth\.edu> writes&#8203;:
> 
> It's the same behavior as nested named subs&#8203;:
> 
> \#\!perl \-w
> 
> sub foo \{
>   my @&#8203;s;
> 
>   sub bar \{
>     push @&#8203;s\, 'x';
>   \}
> 
>   bar\(\);
>   print "$\_\[0\]&#8203;: @&#8203;s\\n";
> \}
> 
> foo\($\_\) for 1\, 2;
> \_\_END\_\_

The above is clearly meaningless because there is no stack frame for foo
at the moment bar is compiled\. So there is no particular @&#8203;s to connect
the lexical var in bar to and you get the standard "variable will not stay
shared"\. In particular\, it's possible to call "bar" even if foo never has
called which shows there's not much the compiler \*can\* do\.

Were it not for this global side effect of setting the function slot of \*bar\,
it could be compiled as effectively&#8203;:

sub foo \{
&nbsp;  my @&#8203;s;

&nbsp;  my $bar = sub \{
&nbsp;    push @&#8203;s\, 'x';
&nbsp;  \}

&nbsp;  $bar\->\(\);
&nbsp;  print "$\_\[0\]&#8203;: @&#8203;s\\n";
\}

\(which would work even if you hide away a $bar closure somewhere and use
that after the foo call finishes\, it will just survive in the foo closure\)

So going back to the original&#8203;:

for \(1\.\.2\) \{
&nbsp;  my @&#8203;s;
&nbsp;  /\(?\{push @&#8203;s\, "x"\}\)/;
&nbsp;  print "$\_&#8203;: @&#8203;s\\n";
\}

You say it's the same\, so I take that to mean that the regex is compiled when
it is seen\, at which point again there is really no activation record yet
and it's supposedly hacked to just use the pad entry for my @&#8203;s that perl
always makes even if you never use or call a piece of code\. Correct so far ?

However\, in this case there is no global reference along which you can 
activate the regex before the activation of the \{\} is entered\. So what stops
it from being handled like a closure \(roughly the way i transformed the
sub bar to my $bar above\) ? That was really what I expected \(?\{\}\) to mean\.
I still don't really see a semantic reason it would need two regex compiles
and/or two code compiles even if the loop is done twice\.
\(though i can very well imagine the current implementation not working
like that\)\. If you agree I think it should be handled as a known
\(low priority\) bug instead of documenting it\.

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

$ perl \-le 'sub a \{ my $p; /\(?\{ print \+\+$p \}\)/ \} a\, a\, a' 
1
2
3

That should print '1' each time\.
This seems to be a very old bug \( see the subject line \)\.

It happens because the \(?\{ \.\. \}\) block uses another pad than the block
it's in&#8203;: so '$p' will have a refcount of 2 because it's used in two
different pads \( see S\_pad\_findlex\(\) in pad\.c \- :752 \)\.

When the 'a' returns for the first time\, the $p lexical will be
'abandonned' instead of 'cleared' \- because its recount is >= 1\.
\( see scope\.c&#8203;:875 \- & run the above script with \-DXv \)\.
So the \(?\{ CODE \}\) will stick to the first instance of $p\.

The PAD\_SAVE\_LOCAL\, PAD\_RESTORE\_LOCAL macros at the EVAL&#8203;: case
in S\_regmatch\(\) aren't at all useful&#8203;: they don't free the lexicals
in the pad\, they only switch pads\.

The obvious solution is to let \(?\{ \.\. \}\) blocks use the pad of the
block they're in; that will also have the advantage of removing that
pad juggling and make things a little bit faster\. 

This doesn't mean of course that
&nbsp;  \{ my $p=1; /\(?\{ my $p = 2 \}\)/; print $p \}
will work differently\.
\(?\{ \.\. \}\) blocks will still be properly scoped\.

The patch draft below passes all tests and works equally well
with precompiled && run\-time compiled regexp's\.

Any issues I missed ?

Note&#8203;:
I think the PAD\_SAVE\_\* macros should be removed completely&#8203;: they're
flawed and anybody trying to use them will eventually fall in the
same trap\.

Regards
Adi



---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/regexec.c	2003-03-09 14:07:31.000000000 +0200
+++ regexec.c	2003-03-16 07:26:12.000000000 +0200
@@ -2832,13 +2832,11 @@ S_regmatch(pTHX_ regnode *prog)
 	    dSP;
 	    OP_4tree *oop = PL_op;
 	    COP *ocurcop = PL_curcop;
-	    PAD *old_comppad;
 	    SV *ret;
 	
 	    n = ARG(scan);
 	    PL_op = (OP_4tree*)PL_regdata->data[n];
 	    DEBUG_r( PerlIO_printf(Perl_debug_log, "  re_eval 0x%"UVxf"\n", PTR2UV(PL_op)) );
-	    PAD_SAVE_LOCAL(old_comppad, (PAD*)PL_regdata->data[n + 2]);
 	    PL_regendp[0] = PL_reg_magic->mg_len = locinput - PL_bostr;
 
 	    {
@@ -2854,7 +2852,6 @@ S_regmatch(pTHX_ regnode *prog)
 	    }
 
 	    PL_op = oop;
-	    PAD_RESTORE_LOCAL(old_comppad);
 	    PL_curcop = ocurcop;
 	    if (logical) {
 		if (logical == 2) {	/* Postponed subexpression. */
--- /arc/perl-current/pp_ctl.c	2003-03-05 01:37:58.000000000 +0200
+++ pp_ctl.c	2003-03-16 05:42:59.000000000 +0200
@@ -2810,10 +2815,10 @@ S_doeval(pTHX_ int gimme, OP** startop, 
 
     /* set up a scratch pad */
 
-    CvPADLIST(PL_compcv) = pad_new(padnew_SAVE);
-
-
-    SAVEMORTALIZESV(PL_compcv);	/* must remain until end of current statement */
+    if (!startop) {
+	CvPADLIST(PL_compcv) = pad_new(padnew_SAVE);
+	SAVEMORTALIZESV(PL_compcv);
+    }
 
     /* make sure we compile in the right package */
 
```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Mar 16\, 2003 at 08&#8203;:00&#8203;:18AM \+0200\, Enache Adrian wrote&#8203;:
> $ perl \-le 'sub a \{ my $p; /\(?\{ print \+\+$p \}\)/ \} a\, a\, a' 
> 1
> 2
> 3
> 
> That should print '1' each time\.
> This seems to be a very old bug \( see the subject line \)\.

It all rather depends on how you view the code within the regex \- whether
it is a block or a sub\. Currently Perl5 treats it as a closure \(in a
half\-baked way\)\, so

&nbsp;    sub a \{ my $x = shift; /\(?\{ $x \}\)/ \}

acts roughly like like

&nbsp;    sub a \{ my $x = shift; sub closure \{ $x \}; /\(?\{ closure\(\) \}\)/ \}

\(In fact with my recent closure patch\, that first code example actually
gives a 'variable will not stay shared' warning\.\)

In the example you give it could be argued that it should be treated
like a block\, but in the case of qr//\, it seems more natural to treat
it as a sub\, since the flow of control is independent of the sub
in which the regex is compiled&#8203;:

&nbsp;    sub a \{ my $x = shift; qr/\(?\{ $x \}\)/ \}
&nbsp;    my $rex = a\(42\);

which seems to be the moral equivalent of

&nbsp;    sub a \{ my $x = shift; sub \{ $x \} \}
&nbsp;    my $sub = a\(42\);

Perhaps bare regexs should be demoted to blocks and qr// should be
promoted to full anon sub status\, so that they get cloned on each execution
of qr// ???

Although I have to say that personally I don't like the idea of sharing
the same pad\, simply becuase of run\-time compilation\, eg

&nbsp;    for \(1\.\.1\_000\_000\) \{
&nbsp;    $p = '\(?\{my $x; \.\.\.\}\)'; /$p/;
&nbsp;    \}

This causes the pad to get filled with a million slots for defunct $x's\.
Eval's have their own pad to avoid this sort of thing\.

> The PAD\_SAVE\_LOCAL\, PAD\_RESTORE\_LOCAL macros at the EVAL&#8203;: case
> in S\_regmatch\(\) aren't at all useful&#8203;: they don't free the lexicals
> in the pad\, they only switch pads\.

I added those two macros; but they were added purely to formalise the
exisitng behaviour of /\(?\{\}\)/\. If we move either to sharing the same pad
or full closure\, they can be removed\.


\-\- 
Lady Nancy Astor&#8203;: If you were my husband\, I would flavour your coffee
with poison\.
Churchill&#8203;: Madam \- if I were your husband\, I would drink it\.

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sun\, Mar 16\, 2003 at 12&#8203;:58&#8203;:19PM \+0000\, Dave Mitchell wrote&#8203;:
> Perhaps bare regexs should be demoted to blocks and qr// should be
> promoted to full anon sub status\, so that they get cloned on each execution
> of qr// ???
> 
> Although I have to say that personally I don't like the idea of sharing
> the same pad\, simply becuase of run\-time compilation\, eg
> 
>     for \(1\.\.1\_000\_000\) \{
>     $p = '\(?\{my $x; \.\.\.\}\)'; /$p/;
>     \}
> 
> This causes the pad to get filled with a million slots for defunct $x's\.
> Eval's have their own pad to avoid this sort of thing\.

I see three cases there \- with their moral equivalent\.

1\. sub a \{ my $x; /\(?\{ $x\+\+ \}\)/ \}
&nbsp;    =>     sub a \{ my $x; \{ $x\+\+ \} \}

2\. sub a \{ my $x; my $p = qr/\(?\{ $x\+\+ \}\)/; $\_ =~ $p \}
&nbsp;    =>    sub a \{ my $x; sub clos \{ $x\+\+ \}; clos \}

3\. sub a \{ my $x; my $p = '\(?\{ $x\+\+ \}\)'; /$p/ \}
&nbsp;    =>    sub a \{ my $x; eval '$x\+\+' \}

Does it look correct to you ?

Regards
Adi

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sun\, Mar 16\, 2003 at 09&#8203;:37&#8203;:51PM \+0200\, Enache Adrian wrote&#8203;:
> 1\. sub a \{ my $x; /\(?\{ $x\+\+ \}\)/ \}
>     =>     sub a \{ my $x; \{ $x\+\+ \} \}
> 
> 2\. sub a \{ my $x; my $p = qr/\(?\{ $x\+\+ \}\)/; $\_ =~ $p \}
>     =>    sub a \{ my $x; sub clos \{ $x\+\+ \}; clos \}
> 
> 3\. sub a \{ my $x; my $p = '\(?\{ $x\+\+ \}\)'; /$p/ \}
>     =>    sub a \{ my $x; eval '$x\+\+' \}

More precisely&#8203;:
3\. sub a \{ my $x; my $p = '\(?\{ $x\+\+ \}\)'; /$p/ \}
&nbsp;    =>      sub a \{ my $x; eval '$x\+\+'; die "\.\.\." if $@&#8203; \}

Revamping all that to use blocks/full closures/full evals will
have another benefic effect&#8203;: get one rid of the top\_env panic
on Perl\_croak\(\)s during the compilation of \(?\{ \.\. \}\) blocks&#8203;:

$ perl \-e '/\(?\{ " \}\)/'
panic&#8203;: top\_env

\(A plethora of tickets are about that\)

Regards
Adi

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Mar 16\, 2003 at 09&#8203;:37&#8203;:51PM \+0200\, Enache Adrian wrote&#8203;:
> On Sun\, Mar 16\, 2003 at 12&#8203;:58&#8203;:19PM \+0000\, Dave Mitchell wrote&#8203;:
> > Perhaps bare regexs should be demoted to blocks and qr// should be
> > promoted to full anon sub status\, so that they get cloned on each execution
> > of qr// ???
> > 
> > Although I have to say that personally I don't like the idea of sharing
> > the same pad\, simply becuase of run\-time compilation\, eg
> > 
> >     for \(1\.\.1\_000\_000\) \{
> >     $p = '\(?\{my $x; \.\.\.\}\)'; /$p/;
> >     \}
> > 
> > This causes the pad to get filled with a million slots for defunct $x's\.
> > Eval's have their own pad to avoid this sort of thing\.
> 
> I see three cases there \- with their moral equivalent\.
> 
> 1\. sub a \{ my $x; /\(?\{ $x\+\+ \}\)/ \}
>     =>     sub a \{ my $x; \{ $x\+\+ \} \}
> 
> 2\. sub a \{ my $x; my $p = qr/\(?\{ $x\+\+ \}\)/; $\_ =~ $p \}
>     =>    sub a \{ my $x; sub clos \{ $x\+\+ \}; clos \}
> 
> 3\. sub a \{ my $x; my $p = '\(?\{ $x\+\+ \}\)'; /$p/ \}
>     =>    sub a \{ my $x; eval '$x\+\+' \}
> 
> Does it look correct to you ?

yes\, except for \(2\)\, which I would put as an anon sub so you get a new
closure for each execution&#8203;:

2\. sub a \{ my $x; my $p = qr/\(?\{ $x\+\+ \}\)/; $\_ =~ $p \}
&nbsp;    =>    sub a \{ my $x; my $p = sub \{ $x\+\+ \}; $p\->\(\) \}

Whether it's practical to implement \(1\)\.\.\(3\)\, and let alone what it would
do to backwards compatibility\, I don't really have a feel for right now\.

I'm also not certain whether it's fair on users to have \(1\) and \(2\)
implemented differently \- that could cause suprises\. eg does

&nbsp;    my $x; $p = 'hello'; /$p\(?\{$x\}\)/

get treated as case \(1\) or \(2\)?

\(Disclaimer&#8203;: I do bug fixes\, not language design\. Hence my general
vagueness here\.\)

\-\- 
"You're so sadly neglected\, and often ignored\.
A poor second to Belgium\, When going abroad\."
Monty Python \- "Finland"

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
